pageserver/src/tenant/mgr.rs

```diff
@@ -36,30 +35,58 @@ use super::delete::DeleteTenantError;
 use super::timeline::delete::DeleteTimelineFlow;
 use super::TenantSharedResources;
 
+/// For a tenant that appears in TenantsMap, it may either be
+/// - `Attached`: has a full Tenant object, is elegible to service
+///    reads and ingest WAL.
+/// - `Secondary`: is only keeping a local cache warm.
+///
+/// Secondary is a totally distinct state rather than being a mode of a `Tenant`, because
+/// that way we avoid having to carefully switch a tenant's ingestion etc on and off during
+/// its lifetime, and we can preserve some important safety invariants like `Tenant` always
+/// having a properly acquired generation (Secondary doesn't need a generation)
+#[derive(Debug)]
+pub enum TenantSlot {
+    Attached(Arc<Tenant>),
+    Secondary,
+    /// In this state, other administrative operations acting on the TenantId should
+    /// block, or return a retry indicator equivalent to HTTP 503
+    InProgress(Arc<tokio::sync::Notify>),
```

Comment:
Could we use the utils::completion::Barrier here instead?

It has a very naive implementation, but if we could agree to use it (more), then I think it would be good.

Reply:
Cut this over to Completion for the RAII convenience of not having to remember to signal the notify :+1: